File: backend/src/main/java/br/ufpr/sept/so2/modules/iam/application/ConsultarSessaoUseCase.java
package br.ufpr.sept.so2.modules.iam.application;

import br.ufpr.sept.so2.modules.iam.application.ports.UsuarioRepository;
import br.ufpr.sept.so2.modules.iam.domain.Usuario;
import br.ufpr.sept.so2.shared.domain.exception.RecursoNaoEncontradoException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

@Service
public class ConsultarSessaoUseCase {

    private final UsuarioRepository usuarioRepository;

    public ConsultarSessaoUseCase(UsuarioRepository usuarioRepository) {
        this.usuarioRepository = usuarioRepository;
    }

    @Transactional(readOnly = true)
    public SessaoAtual execute(UUID usuarioId) {
        Usuario usuario = usuarioRepository.findById(usuarioId)
                .orElseThrow(() -> new RecursoNaoEncontradoException("Usuário não encontrado."));
        return new SessaoAtual(usuario.getId(), usuario.precisaPrimeiroAcesso(), usuario.getAuthorities());
    }

    public record SessaoAtual(UUID id, boolean mustChangePassword, List<String> authorities) {
    }
}

File: backend/src/main/kotlin/br/ufpr/sept/so2/modules/iam/application/ConsultarSessaoUseCase.kt
package br.ufpr.sept.so2.modules.iam.application

import br.ufpr.sept.so2.modules.iam.application.ports.CoordenadorCursosPort
import br.ufpr.sept.so2.modules.iam.application.ports.UsuarioRepository
import br.ufpr.sept.so2.shared.domain.exception.RecursoNaoEncontradoException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.UUID

@Service
class ConsultarSessaoUseCase(
    private val usuarioRepository: UsuarioRepository,
    private val coordenadorCursosPort: CoordenadorCursosPort,
) {
    @Transactional(readOnly = true)
    fun execute(usuarioId: UUID): SessaoAtual {
        val usuario = usuarioRepository.findById(usuarioId)
            .orElseThrow { RecursoNaoEncontradoException("Usuário não encontrado.") }
        val cursoConfigurarId = if (usuario.authorities.contains("course.config")) {
            coordenadorCursosPort.ids(usuario.id).singleOrNull()
        } else {
            null
        }
        return SessaoAtual(
            usuario.id,
            usuario.precisaPrimeiroAcesso(),
            usuario.authorities,
            cursoConfigurarId,
        )
    }

    data class SessaoAtual(
        val id: UUID,
        val mustChangePassword: Boolean,
        val authorities: List<String>,
        val cursoConfigurarId: UUID? = null,
    )
}
